src/main/java/dev/openfga/sdk/api/client/HttpRequestAttempt.java

```diff
@@ -0,0 +1,111 @@
+    private final HttpRequest request;
+
+    public HttpRequestAttempt(
+            HttpRequest request, String name, Class<T> clazz, ApiClient apiClient, BaseConfiguration configuration)
+            throws FgaInvalidParameterException {
+        if (configuration.getMaxRetries() == null) {
+            throw new FgaInvalidParameterException("maxRetries", "Configuration");
+        }
+        this.apiClient = apiClient;
+        this.configuration = configuration;
+        this.name = name;
+        this.request = request;
+        this.clazz = clazz;
+    }
+
+    public CompletableFuture<T> attemptHttpRequest() throws ApiException {
+        int retryNumber = 0;
+        return attemptHttpRequest(apiClient.getHttpClient(), retryNumber, null);
+    }
+
+    private CompletableFuture<T> attemptHttpRequest(HttpClient httpClient, int retryNumber, Throwable previousError) {
+        return httpClient
+                .sendAsync(request, HttpResponse.BodyHandlers.ofString())
+                .thenCompose(response -> {
+                    try {
+                        checkStatus(name, response, previousError);
+                    } catch (FgaApiRateLimitExceededError | FgaApiInternalError retryableError) {
+                        if (retryNumber < configuration.getMaxRetries()) {
+                            HttpClient delayingClient = getDelayedHttpClient();
+                            return attemptHttpRequest(delayingClient, retryNumber + 1, retryableError);
+                        }
+                        return CompletableFuture.failedFuture(retryableError);
+                    } catch (ApiException e) {
+                        return CompletableFuture.failedFuture(e);
+                    }
+
+                    return deserializeResponse(response);
+                });
+    }
+
+    private CompletableFuture<T> deserializeResponse(HttpResponse<String> response) {
+        if (clazz == Void.class && isNullOrWhitespace(response.body())) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        try {
+            T deserialized = apiClient.getObjectMapper().readValue(response.body(), clazz);
+            return CompletableFuture.completedFuture(deserialized);
+        } catch (IOException e) {
+            // Malformed response.
+            return CompletableFuture.failedFuture(new ApiException(e));
+        }
+    }
+
+    private HttpClient getDelayedHttpClient() {
+        Duration retryDelay = configuration.getMinimumRetryDelay();
+        return apiClient
+                .getHttpClientBuilder()
+                .executor(CompletableFuture.delayedExecutor(retryDelay.toNanos(), TimeUnit.NANOSECONDS))
+                .build();
+    }
+
+    private static void checkStatus(String name, HttpResponse<String> response, Throwable previousError)
+            throws ApiException {
+        int status = response.statusCode();
+        String body = response.body();
+
+        switch (status) {
```

Comment:
What is the error thrown when client timeout (i.e., there is no response)?

Reply:
Right now it will end up as a more generic ApiException, it's thrown where I've got the `// Malformed response.` comment

I'm looking into whether I can add a simple timeout test since I don't have any yet

Reply:
Shoot, I don't have a good way to trigger an actual timeout. I would have to get some kind of call where I can mock a body publisher and _not_ mock the HttpClient, and there's not a good way to do this.

Reply:
I should be timing out the request instead of letting the HttpClient timeout though, so making a small change: [56a8101](https://github.com/openfga/sdk-generator/pull/217/commits/56a81016880c1ca9a8eac3c3de6a6a30917f488e)